keras/engine/saving.py

```diff
@@ -16,13 +17,245 @@
 from ..utils.io_utils import ask_to_proceed_with_overwrite
 from ..utils import conv_utils
 
+
+if sys.version_info[0] == 3:
+    import pickle
+else:
+    import cPickle as pickle
+
 try:
     import h5py
     HDF5_OBJECT_HEADER_LIMIT = 64512
 except ImportError:
     h5py = None
 
 
+def _get_json_type(obj):
+    """Serialize any object to a JSON-serializable structure.
+
+    # Arguments
+        obj: the object to serialize
+
+    # Returns
+        JSON-serializable structure representing `obj`.
+
+    # Raises
+        TypeError: if `obj` cannot be serialized.
+    """
+    # if obj is a serializable Keras class instance
+    # e.g. optimizer, layer
+    if hasattr(obj, 'get_config'):
+        return {'class_name': obj.__class__.__name__,
+                'config': obj.get_config()}
+
+    # if obj is any numpy type
+    if type(obj).__module__ == np.__name__:
+        if isinstance(obj, np.ndarray):
+            return {'type': type(obj),
+                    'value': obj.tolist()}
+        else:
+            return obj.item()
+
+    # misc functions (e.g. loss function)
+    if callable(obj):
+        return obj.__name__
+
+    # if obj is a python 'type'
+    if type(obj).__name__ == type.__name__:
+        return obj.__name__
+
+    raise TypeError('Not JSON Serializable:', obj)
+
+
+def get_model_state(model):
```

Comment:
This function would exist exclusively for pickling, so I think it should have "pickle" or some such in the name (e.g. `get_model_state_for_pickling`

Additionally, the code seems highly redundant with save_model / load_model. How can we refactor to minimize code duplication?

Reply:
@fchollet What about - we have a common `get_model_state()`/`load_model_from_state()` which will be called by `pickle_model()`/`unpickle_model()` and `load_model()`/`save_model()`

Reply:
That sounds good!

Reply:
After further thought, that would make h5py write sub-optimal. We would be creating a fully copy of the model weights in memory and writing it all at once to disk. So probably leave `save_model` as is?

Reply:
@fchollet ping ping